### PR TITLE
HttpExchange Kotlin conversion

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

**Comment by yschimke** [comment]
For discussion, the bug in https://github.com/square/okhttp/pull/4936 and the number of !! statements suggests it's a worthy update and ripe for discussion

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+            } else if (statusLine.code == HTTP_CONTINUE) {
+                state = STATE_READ_RESPONSE_HEADERS
+                return responseBuilder
+            }
+
+            state = STATE_OPEN_RESPONSE_BODY
+            return responseBuilder
+        } catch (e: EOFException) {
+            // Provide more context if the server ends the stream before sending a response.
+            val address = if (realConnection != null) realConnection.route().address().url().redact() else "unknown"
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: TODO: confirm logic for fix from https://github.com/square/okhttp/pull/4936

File: okhttp/src/main/java/okhttp3/internal/http/ExchangeCode.kt [hunk 2]
```diff
@@ -27,7 +27,7 @@ import okio.Source
 interface ExchangeCodec {
 
   /** Returns the connection that carries this codec.  */
-  fun connection(): RealConnection
+  fun connection(): RealConnection?
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http/ExchangeCode.kt]: mind renaming this file to `ExchangeCodec.kt`?

File: okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    private val protocol: Protocol = if (client.protocols().contains(Protocol.H2_PRIOR_KNOWLEDGE))
+        Protocol.H2_PRIOR_KNOWLEDGE
+    else
+        Protocol.HTTP_2
+
+    @Volatile
+    private var canceled: Boolean = false
+
+    override fun connection(): RealConnection {
+        return realConnection
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt]: +2 indents?

**Comment by swankjesse** [comment]
Side-by-side diff? https://gist.github.com/jrodbx/dcc81cb213e4c300c0e1ddfde4ae553d

**Comment by yschimke** [comment]
> Side-by-side diff? https://gist.github.com/jrodbx/dcc81cb213e4c300c0e1ddfde4ae553d

https://github.com/square/okhttp/pull/4938/commits/dbadaa6b614d02b37b6c7c773a418fd0216d85d6

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 4]
```diff
@@ -13,530 +13,542 @@
...
- *     <li>Write to and then close that sink.
- *     <li>{@linkplain #readResponseHeaders Read response headers}.
- *     <li>Open a source to read the response body. Either {@linkplain #newFixedLengthSource
- *         fixed-length}, {@linkplain #newChunkedSource chunked} or {@linkplain
- *         #newUnknownLengthSource unknown length}.
- *     <li>Read from and close that source.
- * </ol>
+ *  1. [Send request headers][.writeRequest].
+ *  2. Open a sink to write the request body. Either [known][newKnownLengthSink] or
+ *  [chunked][newChunkedSink].
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: nit: indent wrapped to line up with `Open...`

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 5]
```diff
@@ -13,530 +13,542 @@
...
- *     <li>Open a sink to write the request body. Either {@linkplain #newKnownLengthSink known
- *         length} or {@link #newChunkedSink chunked}.
- *     <li>Write to and then close that sink.
- *     <li>{@linkplain #readResponseHeaders Read response headers}.
- *     <li>Open a source to read the response body. Either {@linkplain #newFixedLengthSource
- *         fixed-length}, {@linkplain #newChunkedSource chunked} or {@linkplain
- *         #newUnknownLengthSource unknown length}.
- *     <li>Read from and close that source.
- * </ol>
+ *  1. [Send request headers][.writeRequest].
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: drop the `.` in `[.writeRequest]`. .. . the Kotlin conversion adds it and it doesn’t work

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: (here and elsewhere)

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 6]
```diff
@@ -13,530 +13,542 @@
...
-    }
+class Http1ExchangeCodec(
+  /** The client that configures this stream. May be null for HTTPS proxy tunnels.  */
+  private val client: OkHttpClient?,
+  /** The connection that carries this stream.  */
+  private val realConnection: RealConnection?,
+  private val source: BufferedSource,
+  private val sink: BufferedSink
+) : ExchangeCodec {
+    private var state = STATE_IDLE
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: Indentation is wrong here

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 7]
```diff
@@ -13,530 +13,542 @@
...
+    override fun cancel() {
+        realConnection?.cancel()
     }
 
-    long contentLength = HttpHeaders.contentLength(response);
-    if (contentLength != -1) {
-      return newFixedLengthSource(contentLength);
+    /**
+     * Prepares the HTTP headers and sends them to the server.
+     *
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: nit: remove extra blank lines

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 8]
```diff
@@ -13,530 +13,542 @@
...
+        if (!HttpHeaders.hasBody(response)) {
+            return 0L
+        }
 
-  @Override public Headers trailers() {
-    if (state != STATE_CLOSED) {
-      throw new IllegalStateException("too early; can't read the trailers yet");
+        return if ("chunked".equals(response.header("Transfer-Encoding"), ignoreCase = true)) {
+            -1L
+        } else HttpHeaders.contentLength(response)
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: curlies

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 9]
```diff
@@ -13,530 +13,542 @@
...
-    for (String line; (line = readHeaderLine()).length() != 0; ) {
-      addHeaderLenient(headers, line);
+        if ("chunked".equals(response.header("Transfer-Encoding"), ignoreCase = true)) {
+            return newChunkedSource(response.request().url())
+        }
+
+        val contentLength = HttpHeaders.contentLength(response)
+        return if (contentLength != -1L) {
+            newFixedLengthSource(contentLength)
+        } else newUnknownLengthSource()
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: curlies

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 10]
```diff
@@ -13,530 +13,542 @@
...
-  /** An HTTP request body. */
-  private final class KnownLengthSink implements Sink {
-    private final ForwardingTimeout timeout = new ForwardingTimeout(sink.timeout());
-    private boolean closed;
-
-    @Override public Timeout timeout() {
-      return timeout;
+
+    override fun trailers(): Headers {
+        if (state != STATE_CLOSED) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]:
```
check(state == STATE_CLOSED) { "too early; can't read the trailers yet" }
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 11]
```diff
@@ -13,530 +13,542 @@
...
 
-    @Override public void close() throws IOException {
-      if (closed) return;
-      closed = true;
-      detachTimeout(timeout);
-      state = STATE_READ_RESPONSE_HEADERS;
+    /** Returns bytes of a request header for sending on an HTTP transport.  */
+    @Throws(IOException::class)
+    fun writeRequest(headers: Headers, requestLine: String) {
+        if (state != STATE_IDLE) throw IllegalStateException("state: $state")
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]:
```
check(sate == STATE_IDLE)
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 12]
```diff
@@ -13,530 +13,542 @@
...
+        sink.writeUtf8(requestLine).writeUtf8("\r\n")
+        var i = 0
+        val size = headers.size()
+        while (i < size) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]:
```
for (i in 0 until headers.size()) {
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 13]
```diff
@@ -13,530 +13,542 @@
...
-  /**
-   * An HTTP body with alternating chunk sizes and chunk bodies. It is the caller's responsibility
-   * to buffer chunks; typically by using a buffered sink with this sink.
-   */
-  private final class ChunkedSink implements Sink {
-    private final ForwardingTimeout timeout = new ForwardingTimeout(sink.timeout());
-    private boolean closed;
+    @Throws(IOException::class)
+    override fun readResponseHeaders(expectContinue: Boolean): Response.Builder? {
+        if (state != STATE_OPEN_REQUEST_BODY && state != STATE_READ_RESPONSE_HEADERS) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: check

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 14]
```diff
@@ -13,530 +13,542 @@
...
+        if (state != STATE_OPEN_REQUEST_BODY) throw IllegalStateException("state: $state")
+        state = STATE_WRITING_REQUEST_BODY
+        return ChunkedSink()
     }
 
-    @Override public synchronized void flush() throws IOException {
-      if (closed) return; // Don't throw; this stream might have been closed on the caller's behalf.
-      sink.flush();
+    private fun newKnownLengthSink(): Sink {
+        if (state != STATE_OPEN_REQUEST_BODY) throw IllegalStateException("state: $state")
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: `check(...)` here and elsewhere

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 15]
```diff
@@ -13,530 +13,542 @@
...
+            }
+        }
+
+        @Throws(IOException::class)
+        override fun read(sink: Buffer, byteCount: Long): Long {
+            if (byteCount < 0) throw IllegalArgumentException("byteCount < 0: $byteCount")
+            if (closed) throw IllegalStateException("closed")
+            if (bytesRemaining == 0L) return -1
+
+            val read = super.read(sink, Math.min(bytesRemaining, byteCount))
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: `minOf()`

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 16]
```diff
@@ -13,530 +13,542 @@
...
+            }
+            return read
+        }
+
+        @Throws(IOException::class)
+        override fun close() {
+            if (closed) return
+
+            if (bytesRemaining != 0L && !Util.discard(this,
+                            ExchangeCodec.DISCARD_STREAM_TIMEOUT_MILLIS, MILLISECONDS)) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: this line wrapping is 🍌

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java [hunk 17]
```diff
@@ -13,530 +13,542 @@
...
+            if (bytesRemainingInChunk != NO_CHUNK_YET) {
+                source.readUtf8LineStrict()
+            }
+            try {
+                bytesRemainingInChunk = source.readHexadecimalUnsignedLong()
+                val extensions = source.readUtf8LineStrict().trim { it <= ' ' }
+                if (bytesRemainingInChunk < 0 || extensions.isNotEmpty() && !extensions.startsWith(
+                                ";")) {
+                    throw ProtocolException(
+                            "expected chunk size and optional extensions but was \"" +
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: template string

File: okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.java [hunk 18]
```diff
@@ -13,194 +13,204 @@
...
+  private val chain: Interceptor.Chain,
+  private val connection: Http2Connection
+) :
+        ExchangeCodec {
+    @Volatile
+    private var stream: Http2Stream? = null
+
+    private val protocol: Protocol = if (client.protocols().contains(Protocol.H2_PRIOR_KNOWLEDGE))
+        Protocol.H2_PRIOR_KNOWLEDGE
+    else
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.java]: curlies. (Whenever lines wrap, curlies!)

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java (hunk 14 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: Think we need some inlinable conversion for this, made a trivial mistake while flipping the bits.  Easier to retain, inline, then simplify with Intellij.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java]: I add an empty else clause, then use “invert if” and I have all the missing pieces I need.

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 19]
```diff
@@ -0,0 +1,527 @@
...
+      HttpHeaders.contentLength(response)
+    }
+  }
+
+  override fun openResponseBodySource(response: Response): Source {
+    if (!HttpHeaders.hasBody(response)) {
+      return newFixedLengthSource(0)
+    }
+
+    if ("chunked".equals(response.header("Transfer-Encoding"), ignoreCase = true)) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: Tempted to have a private `isChunked` extension for `Response`

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 20]
```diff
@@ -0,0 +1,527 @@
...
+    sink.flush()
+  }
+
+  /** Returns bytes of a request header for sending on an HTTP transport.  */
+  fun writeRequest(headers: Headers, requestLine: String) {
+    check(state == STATE_IDLE) { "state: $state" }
+    sink.writeUtf8(requestLine).writeUtf8("\r\n")
+    var i = 0
+    val size = headers.size()
+    while (i < size) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]:
```kotlin
for (i in 0 until headers.size()) {
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 21]
```diff
@@ -0,0 +1,527 @@
...
+
+  /**
+   * The response body from a CONNECT should be empty, but if it is not then we should consume it
+   * before proceeding.
+   */
+  fun skipConnectBody(response: Response) {
+    val contentLength = HttpHeaders.contentLength(response)
+    if (contentLength == -1L) return
+    val body = newFixedLengthSource(contentLength)
+    Util.skipAll(body, Integer.MAX_VALUE, MILLISECONDS)
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]:
```suggestion
    Util.skipAll(body, Int.MAX_VALUE, MILLISECONDS)
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 22]
```diff
@@ -0,0 +1,527 @@
...
+      detachTimeout(timeout)
+      state = STATE_READ_RESPONSE_HEADERS
+    }
+  }
+
+  /**
+   * An HTTP body with alternating chunk sizes and chunk bodies. It is the caller's responsibility
+   * to buffer chunks; typically by using a buffered sink with this sink.
+   */
+  private inner class ChunkedSink internal constructor() : Sink {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: Do you need `internal constructor`?

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 23]
```diff
@@ -0,0 +1,527 @@
...
+      AbstractSource() {
+
+    init {
+      if (bytesRemaining == 0L) {
+        responseBodyComplete()
+      }
+    }
+
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      if (byteCount < 0) throw IllegalArgumentException("byteCount < 0: $byteCount")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]:
```suggestion
      require(byteCount >= 0) { "byteCount < 0: $byteCount" }
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 24]
```diff
@@ -0,0 +1,527 @@
...
+      if (closed) throw IllegalStateException("closed")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]:
```suggestion
     check(!closed) { "closed" }
```

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 25]
```diff
@@ -0,0 +1,527 @@
...
+      if (bytesRemaining == 0L) return -1
+
+      val read = super.read(sink, minOf(bytesRemaining, byteCount))
+      if (read == -1L) {
+        realConnection!!.noNewExchanges() // The server didn't supply the promised content length.
+        val e = ProtocolException("unexpected end of stream")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: Inline

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 26]
```diff
@@ -0,0 +1,527 @@
...
+
+  /** An HTTP body with alternating chunk sizes and chunk bodies.  */
+  private inner class ChunkedSource internal constructor(private val url: HttpUrl) :
+      AbstractSource() {
+    private var bytesRemainingInChunk = NO_CHUNK_YET
+    private var hasMoreChunks = true
+
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      if (byteCount < 0) throw IllegalArgumentException("byteCount < 0: $byteCount")
+      if (closed) throw IllegalStateException("closed")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: require + check

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 27]
```diff
@@ -0,0 +1,527 @@
...
+
+      if (bytesRemainingInChunk == 0L || bytesRemainingInChunk == NO_CHUNK_YET) {
+        readChunkSize()
+        if (!hasMoreChunks) return -1
+      }
+
+      val read = super.read(sink, minOf(byteCount, bytesRemainingInChunk))
+      if (read == -1L) {
+        realConnection!!.noNewExchanges() // The server didn't supply the promised chunk length.
+        val e = ProtocolException("unexpected end of stream")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: inline

File: okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt [hunk 28]
```diff
@@ -0,0 +1,527 @@
...
+    }
+  }
+
+  /** An HTTP message body terminated by the end of the underlying stream.  */
+  private inner class UnknownLengthSource : AbstractSource() {
+    private var inputExhausted: Boolean = false
+
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      if (byteCount < 0) throw IllegalArgumentException("byteCount < 0: $byteCount")
+      if (closed) throw IllegalStateException("closed")
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt]: require + check

File: okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt [hunk 29]
```diff
@@ -0,0 +1,211 @@
...
+/** Encode requests and responses using HTTP/2 frames.  */
+class Http2ExchangeCodec(
+  client: OkHttpClient,
+  private val realConnection: RealConnection,
+  private val chain: Interceptor.Chain,
+  private val connection: Http2Connection
+) : ExchangeCodec {
+  @Volatile private var stream: Http2Stream? = null
+
+  private val protocol: Protocol = if (client.protocols().contains(Protocol.H2_PRIOR_KNOWLEDGE)) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt]:
```suggestion
  private val protocol: Protocol = if (Protocol.H2_PRIOR_KNOWLEDGE in client.protocols()) {
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+        result.add(Header(TARGET_AUTHORITY, host)) // Optional.
+      }
+      result.add(Header(TARGET_SCHEME, request.url().scheme()))
+
+      var i = 0
+      val size = headers.size()
+      while (i < size) {
+        // header names must be lowercase.
+        val name = headers.name(i).toLowerCase(Locale.US)
+        if (!HTTP_2_SKIPPED_REQUEST_HEADERS.contains(name) || name == TE && headers.value(
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt]:
```suggestion
        if (name !in HTTP_2_SKIPPED_REQUEST_HEADERS || name == TE && headers.value(
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt [hunk 31]
```diff
@@ -0,0 +1,211 @@
...
+      result.add(Header(TARGET_PATH, RequestLine.requestPath(request.url())))
+      val host = request.header("Host")
+      if (host != null) {
+        result.add(Header(TARGET_AUTHORITY, host)) // Optional.
+      }
+      result.add(Header(TARGET_SCHEME, request.url().scheme()))
+
+      var i = 0
+      val size = headers.size()
+      while (i < size) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt]:
```kotlin
for (i in 0 until headers.size())
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt [hunk 32]
```diff
@@ -0,0 +1,211 @@
...
+      return result
+    }
+
+    /** Returns headers for a name value block containing an HTTP/2 response.  */
+    fun readHttp2HeadersList(headerBlock: Headers, protocol: Protocol): Response.Builder {
+      var statusLine: StatusLine? = null
+      val headersBuilder = Headers.Builder()
+      var i = 0
+      val size = headerBlock.size()
+      while (i < size) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt]:
```kotlin
for (i in 0 until headerBlock.size()) {
```

**Comment by yschimke** [comment]
Thanks for the reviews, this was a good learning experience for the migration and the kotlin is so much nicer.

**Comment by yschimke** [comment]
Failing on https://travis-ci.org/square/okhttp/jobs/519834499

```
> Task :okhttp:compileKotlin
e: /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt: (31, 41): Unresolved reference: HTTP_CONTINUE
e: /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt: (187, 46): Unresolved reference: HTTP_CONTINUE
e: /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.kt: (188, 28): Unresolved reference: HTTP_CONTINUE
e: /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt: (32, 41): Unresolved reference: HTTP_CONTINUE
e: /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/internal/http2/Http2ExchangeCodec.kt: (103, 77): Unresolved reference: HTTP_CONTINUE
```

**Pull request closed without merging**